### PR TITLE
added tests for Array.prototype.group

### PR DESCRIPTION
**Pull request by polsevev**
Added tests for proposal Array Grouping groupBy

CC @codehag

### CODE REVIEW

**Comment by rwaldron** [comment]
Again, very cool that you jumped right in on these :D we'll also go over this one during our video call on Friday. See you then!!

**Review by ljharb — COMMENTED** [review]
This looks great! I compared it against my polyfill's tests, and noticed this PR isn't testing that a non-function callbackfn throws a TypeError, and that the [[Prototype]] of the returned object is null.

File: test/built-ins/Array/prototype/groupBy/callback-arg.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  6. Repeat, while k < len
+
+      a. Let Pk be ! ToString(𝔽(k)).
+      b. Let kValue be ? Get(O, Pk).
+      c. Let propertyKey be ? ToPropertyKey(? Call(callbackfn, thisArg, « kValue, 𝔽(k), O »)).
+      d. Perform ! AddValueToKeyedGroup(groups, propertyKey, kValue).
+      e. Set k to k + 1.
+  ...
+
```
Comment (rwaldron) [inline test/built-ins/Array/prototype/groupBy/callback-arg.js:20]:
```suggestion
features: [Array.prototype.groupBy]
```

File: test/built-ins/Array/prototype/groupBy/null.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+
+  ...
+
+---*/
+
+
+assert.throws(TypeError, function() {
+  Array.prototype.groupBy.call(null,null);
```
Comment (rwaldron) [inline test/built-ins/Array/prototype/groupBy/null.js:20]:
```suggestion
Array.prototype.groupBy.call(null, () => {
  throw new Test262Error();
});
```

Comment (rwaldron) [reply test/built-ins/Array/prototype/groupBy/null.js:20]: And combine all of the invalid this tests into one file :)

File: test/built-ins/Array/prototype/groupBy/callback-arg.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+---*/
+
+
+const arr = [-0,0,1,2,3];
+
+const sentinel = {};
+
+const map = arr.groupBy.call(sentinel, (n,i,testArr) => {
```
Comment (legendecas) [inline test/built-ins/Array/prototype/groupBy/callback-arg.js:28]: IIUC, `LengthOfArrayLike` of the `sentinel` is 0, the callback won't be called in this case.

File: test/built-ins/Array/prototype/groupBy/length.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+assert.sameValue(
+  Array.prototype.groupBy.length, 1,
+  'The value of `Array.prototype.groupBy.length` is `1`'
+);
+  
+verifyProperty(Array.prototype.groupBy, "length", {
+  enumerable: false,
+  writable: false,
+  configurable: true
+});
```
Comment (leobalter) [inline test/built-ins/Array/prototype/groupBy/length.js:29]:
```suggestion
verifyProperty(Array.prototype.groupBy, "length", {
  value: 1,
  enumerable: false,
  writable: false,
  configurable: true
});
```

File: test/built-ins/Array/prototype/groupBy/name.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+assert.sameValue(
+    Array.prototype.groupBy.name, 'groupBy',
+    'The value of `Array.prototype.groupBy.name` is `"groupBy"`'
+);
+  
+verifyProperty(Array.prototype.groupBy, "name", {
+  enumerable: false,
+  writable: false,
+  configurable: true
+});
```
Comment (leobalter) [inline test/built-ins/Array/prototype/groupBy/name.js:28]:
```suggestion
verifyProperty(Array.prototype.groupBy, "name", {
  value: "groupBy",
  enumerable: false,
  writable: false,
  configurable: true
});
```

File: test/built-ins/Array/prototype/groupBy/undefined.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+
+  ...
+
+---*/
+
+
+assert.throws(TypeError, function() {
+  Array.prototype.groupBy.call(undefined);
+});
```
Comment (leobalter) [inline test/built-ins/Array/prototype/groupBy/undefined.js:21]:
```suggestion
var calls = 0;

assert.throws(TypeError, function() {
  Array.prototype.groupBy.call(undefined, () => calls++);
});
assert.sameValue(calls, 0, "callbackfn should not be called");
```

File: test/built-ins/Array/prototype/groupBy/emptyList.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    a. Let elements be ! CreateArrayFromList(g.[[Elements]]).
+    b. Perform ! CreateDataPropertyOrThrow(obj, g.[[Key]], elements).
+
+  ...
+
+---*/
+
+const obj = [].groupBy(i=>i);
+
+assert.sameValue(Object.keys(obj).length, 0);
```
Comment (leobalter) [inline test/built-ins/Array/prototype/groupBy/emptyList.js:23]: You might also benefit testing that `obj` is different to the original array and asserting the callbackfn is not called. Something like:

```js
var calls = 0;
const original = [];
const obj = original.groupBy(() => calls++);

assert.sameValue(calls, 0, "callback function should not be called");
assert.notSameValue(original, obj, "groupBy returns a different object");
assert.sameValue(Object.keys(obj).length, 0);
```

**Review by leobalter — CHANGES_REQUESTED** [review]
The tests are good, I added a few suggestions only.

You may also wanna test:

- the return type of groupBy
- assert this value inside the callbackfn w/ and w/o `thisArg`
- non-callback values for callbackfn
- non valid values for the this value
- tainted callbackfns (throwing an exception).
- callbackfn calls with terse arrays like `[,,,1,,2]`
- array like values for the this value, e.g. `[].groupBy.call({length: 2, '1': 'foo'}, callbackfn);`

**Comment by hemanth** [comment]
This should be `group` instead of `groupBy` now.

**Review by jridgewell — COMMENTED** [review]
https://github.com/tc39/test262/pull/3354#pullrequestreview-848251319

> the return type of groupBy

done.

> assert this value inside the callbackfn w/ and w/o thisArg

done.

> non-callback values for callbackfn

done.

> non valid values for the this value

done.

> tainted callbackfns (throwing an exception).

Do we have to?

> callbackfn calls with terse arrays like [,,,1,,2]

done.

> array like values for the this value, e.g. [].groupBy.call({length: 2, '1': 'foo'}, callbackfn);

Done.

File: test/built-ins/Array/prototype/group/callback-arg.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      e. Set k to k + 1.
+  ...
+features: [Array.prototype.group]
+---*/
+
+
+const arr = [-0, 0, 1, 2, 3];
+
+const sentinel = {};
+let calls = 0;
```
Comment (ptomato) [inline test/built-ins/Array/prototype/group/callback-arg.js:27]: `sentinel` is not used, this seems to be left over from one of the other tests?
```suggestion
let calls = 0;
```

File: test/built-ins/Array/prototype/group/emptyList.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021 Ecma International.  All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-array.prototype.group
+description: Array.prototype.group populates object with correct keys and values
```
Comment (ptomato) [inline test/built-ins/Array/prototype/group/emptyList.js:6]: This description doesn't quite seem to fit the test; maybe something like this?
```suggestion
description: Callback is not called and object is not populated if the array is empty
```

File: test/built-ins/Array/prototype/group/groupLength.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021 Ecma International.  All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-array.prototype.group
+description: Array.prototype.group populates object with correct keys and values
```
Comment (ptomato) [inline test/built-ins/Array/prototype/group/groupLength.js:6]: Same here, I'd suggest something like:
```suggestion
description: Callback can return numbers that are converted to property keys
```

File: test/built-ins/Array/prototype/group/sparse-array.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+const obj = array.group(() => {
+  calls++;
+  return 'key';
+});
+
+assert.sameValue(calls, 3);
+
+assert.sameValue(0 in obj['key'], true);
+assert.sameValue(1 in obj['key'], true);
+assert.sameValue(2 in obj['key'], true);
```
Comment (ptomato) [inline test/built-ins/Array/prototype/group/sparse-array.js:35]: `assert.sameValue(..., true)` is going to produce a pretty context-less error message if it fails, something like "Expected true but got false". Plain `assert()` won't be any better (at least it's shorter) but I'd recommend adding assertion messages here:
```suggestion
assert(0 in obj['key'], 'group has a first element');
assert(1 in obj['key'], 'group has a second element');
assert(2 in obj['key'], 'group has a third element');
```

File: test/built-ins/Array/prototype/group/sparse-array.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2021 Ecma International.  All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-array.prototype.group
+description: Array.prototype.group populates object with correct keys and values
```
Comment (ptomato) [inline test/built-ins/Array/prototype/group/sparse-array.js:6]:
```suggestion
description: Array.prototype.group adds undefined to the group for sparse arrays
```

**Review by ptomato — APPROVED** [review]
Thanks for resuming this PR, and apologies that it took some time to get around to it. (I'll start on groupToMap as soon as possible.) I think this is basically ready to merge. Some small suggestions, but keeping in mind how long this has already taken to review, I've made them into GitHub suggestions that you can just click to accept or not, so that the review process don't cause any more delay on this PR.

I have some suggestions for additional things that could be tested, but for the same reason I wouldn't say that they should block this PR. They can be added in a second PR if you are so inclined:
- Test throwing on an arraylike with a value of its `length` property that can't be converted to an integer, or a getter that throws
- Test throwing in step 6.b if _Pk_ is a property with a getter that throws
- Test throwing if the callback returns a value that can't be converted to a property key (e.g. an object with a toString method that throws)

Thanks again for resuming this work!

**Comment by ptomato** [comment]
> > tainted callbackfns (throwing an exception).
> 
> Do we have to?

Would be nice, but as with the other suggestions I don't see a need to block this PR on having that.

Another one I thought of: shortening and lengthening the array during the iteration.

**Comment by jridgewell** [comment]
> Test throwing on an arraylike with a value of its length property that can't be converted to an integer, or a getter that throws

Done.

> Test throwing in step 6.b if Pk is a property with a getter that throws

Done.

> Test throwing if the callback returns a value that can't be converted to a property key (e.g. an object with a toString method that throws)

Done.

> tainted callbackfns (throwing an exception).

Done.

**Review by ptomato — APPROVED** [review]
Awesome, thanks!

**Pull request merged**